File: app/src/main/java/com/quran/labs/androidquran/dao/translation/TranslationItem.java
package com.quran.labs.androidquran.dao.translation;

public class TranslationItem implements TranslationRowData {
  public final int localVersion;
  public final Translation translation;

  public TranslationItem(Translation translation) {
    this(translation, 0);
  }

  public TranslationItem(Translation translation, int localVersion) {
    this.translation = translation;
    this.localVersion = localVersion;
  }

  public boolean exists() {
    return localVersion > 0;
  }

  @Override
  public String name() {
    return this.translation.displayName;
  }

  @Override
  public boolean isSeparator() {
    return false;
  }

  @Override
  public boolean needsUpgrade() {
    return localVersion > 0 && this.translation.currentVersion > this.localVersion;
  }

  public TranslationItem withTranslationRemoved() {
    return new TranslationItem(this.translation, 0);
  }

  public TranslationItem withTranslationVersion(int version) {
    return new TranslationItem(this.translation, version);
  }
}

File: app/src/main/java/com/quran/labs/androidquran/dao/translation/TranslationItem.kt
package com.quran.labs.androidquran.dao.translation

import com.quran.mobile.translation.model.LocalTranslation

data class TranslationItem @JvmOverloads constructor(val translation: Translation,
                                                     val localVersion: Int = 0,
                                                     val displayOrder: Int = -1) : TranslationRowData {

  override fun isSeparator() = false

  fun exists() = localVersion > 0

  override fun name() = this.translation.displayName

  override fun needsUpgrade(): Boolean {
    return localVersion > 0 && this.translation.currentVersion > this.localVersion
  }

  fun withTranslationRemoved() = this.copy(localVersion = 0)

  fun withDisplayOrder(newDisplayOrder: Int) = this.copy(displayOrder = newDisplayOrder)

  fun withLocalVersionAndDisplayOrder(newVersion: Int, displayOrder: Int) = this.copy(localVersion = newVersion, displayOrder = displayOrder)

  fun asLocalTranslation(): LocalTranslation {
    return LocalTranslation(
      id = translation.id.toLong(),
      filename = translation.fileName,
      name = translation.displayName,
      translator = translation.translator,
      translatorForeign = translation.translatorNameLocalized,
      url = translation.fileUrl,
      languageCode = translation.languageCode,
      version = localVersion,
      minimumVersion = translation.minimumVersion,
      displayOrder = displayOrder
    )
  }
}
